apps/wallet/src/ui/app/hooks/deepbook/useBalanceConversion.ts

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+import { Coins, useDeepBookConfigs } from '_app/hooks';
+import { DEEPBOOK_KEY } from '_pages/swap/constants';
+import { useDeepBookContext } from '_shared/deepBook/context';
+import { type DeepBookClient } from '@mysten/deepbook';
+import { useQuery } from '@tanstack/react-query';
+import BigNumber from 'bignumber.js';
+import { useMemo } from 'react';
+
+async function getDeepBookPriceForCoin(
+	coin: Coins,
+	pools: Record<string, string[]>,
+	isAsk: boolean,
+	deepBookClient: DeepBookClient,
+) {
+	if (coin === Coins.USDC) {
+		return 1n;
+	}
+
+	const poolName = `${coin}_USDC`;
+	const poolIds = pools[poolName];
+	const promises = poolIds.map(async (poolId) => {
+		const { bestBidPrice, bestAskPrice } = await deepBookClient.getMarketPrice(poolId);
+
+		return isAsk ? bestBidPrice : bestAskPrice;
+	});
+
+	const prices = await Promise.all(promises);
+
+	const filter: bigint[] = prices.filter((price): price is bigint => {
+		return typeof price === 'bigint' && price !== 0n;
+	});
+
+	const total = filter.reduce((acc, price) => {
+		return acc + price;
+	}, 0n);
+
+	return total / BigInt(filter.length);
+}
+
+function useDeepbookPricesInUSD(coins: Coins[], isAsk: boolean) {
+	const deepBookClient = useDeepBookContext().client;
+	const deepbookPools = useDeepBookConfigs().pools;
+
+	return useQuery({
+		queryKey: [DEEPBOOK_KEY, 'get-prices-usd', coins, isAsk],
+		queryFn: async () => {
+			const promises = coins.map((coin) =>
+				getDeepBookPriceForCoin(coin, deepbookPools, isAsk, deepBookClient),
+			);
+			return Promise.all(promises);
+		},
+	});
+}
+
+function useAveragePrice(base: Coins, quote: Coins, isAsk: boolean) {
+	const { data: prices, ...rest } = useDeepbookPricesInUSD([base, quote], isAsk);
+
+	const averagePrice = useMemo(() => {
+		const basePrice = new BigNumber((prices?.[0] ?? 1n).toString());
+		const quotePrice = new BigNumber((prices?.[1] ?? 1n).toString());
+
+		const basePriceBigNumber = new BigNumber(basePrice.toString());
+		const quotePriceBigNumber = new BigNumber(quotePrice.toString());
+
+		let avgPrice;
+		if (quote === Coins.USDC) {
+			avgPrice = basePriceBigNumber;
+		} else {
+			avgPrice = basePriceBigNumber.dividedBy(quotePriceBigNumber);
+		}
+
+		return avgPrice;
+	}, [prices, quote]);
+
+	return {
+		data: averagePrice,
+		...rest,
+	};
+}
+
+export function useBalanceConversion(
+	balance: BigInt | BigNumber | null,
+	from: Coins,
+	to: Coins,
+	conversionRate: number = 1,
+) {
+	const { data: averagePrice, ...rest } = useAveragePrice(from, to, to === Coins.USDC);
```

Comment:
you could probs also use `select` here similar to the previous comment to transform `averagePrice` -> ` { rawValue, averagePriceWithConversion }` but eh will leave that to you

I'd probably remove the `useMemo` again unless that calculation is expensive (e.g. you do `console.time()` and then `console.timeEnd()` after the operation and see if it takes a good while)